Allow setting the build output image reference in oc new-build

rhcarvalho opened the pull request:
This makes it possible to do:
```
oc new-build <source> --to=somenamespace/some:other
```

The above will define a build output image and works in concert with `--to-docker` (and `--name`, too).

Fixes #4980.

rhcarvalho commented:
This alone does not make it possible to have no output for a build.

@smarterclayton do you think we should have another argument like `--no-output` or a special value of `--to`?
AFAIK with `cobra`/`pflag` there's no obvious way to determine if a flag was user provided or defaulted.

rhcarvalho commented:
There was an unreported bug with `--to-docker` that is being fixed here, as I noted in the PR's description.

rhcarvalho commented:
[test]

rhcarvalho commented:
@smarterclayton PTAL

smarterclayton commented on pkg/generate/app/app.go:
```diff
@@ -202,6 +202,37 @@ type ImageRef struct {
+func NewOutputImageRef(source *SourceRef, input *ImageRef, to string, outputDocker bool) (*ImageRef, error) {
+	var outputImageRef imageapi.DockerImageReference
+
+	if to != "" {
```
We always use `len(to) > 0` or `len(to) == 0` for this.

rhcarvalho replied:
Updated that.

smarterclayton commented on pkg/generate/app/cmd/newapp.go:
```diff
@@ -542,8 +543,36 @@ func (c *AppConfig) buildPipelines(components app.ComponentReferences, environme
+				from := ref.Input().String()
+				env := c.GetBuildEnvironment(environment)
+
+				output, err := app.NewOutputImageRef(source, input, c.To, c.OutputDocker)
```
Extract this into a method, it's too big to be here.

rhcarvalho replied:
I've broken `buildPipelines` down into smaller pieces.

smarterclayton commented on pkg/generate/app/cmd/newapp.go:
```diff
@@ -542,8 +543,36 @@ func (c *AppConfig) buildPipelines(components app.ComponentReferences, environme
+				// (via --to in the CLI), just print a warning and proceed.
+				// We need to apply defaults to the Docker image reference, so that e.g.
+				// "foo/bar" equals "foo/bar:latest", etc.
+				defaultedInputRef := input.DockerImageReference.DockerClientDefaults()
```
You should implement an equality method in image/api/helpers.go and use that

rhcarvalho replied:
Done.

smarterclayton commented on pkg/generate/app/cmd/newapp.go:
```diff
@@ -542,8 +543,36 @@ func (c *AppConfig) buildPipelines(components app.ComponentReferences, environme
 				}
-				if pipeline, err = app.NewBuildPipeline(ref.Input().String(), input, c.OutputDocker, strategy, c.GetBuildEnvironment(environment), source); err != nil {
-					return nil, fmt.Errorf("can't build %q: %v", ref.Input(), err)
+
```
It's illegal to specify --to when there are more than two builds defined - that check needs to be represented somewhere.

rhcarvalho replied:
is it the same for --name? What kind of use of new-app /new-build produces 2 builds?

smarterclayton replied:
You can specify multiple pipelines today.

On Oct 19, 2015, at 8:38 AM, Rodolfo Carvalho <notifications@github.com>
wrote:

In pkg/generate/app/cmd/newapp.go
<https://github.com/openshift/origin/pull/5141#discussion_r42365386>:

> @@ -542,8 +543,36 @@ func (c *AppConfig) buildPipelines(components app.ComponentReferences, environme
>  						}
>  					}
>  				}
> -				if pipeline, err = app.NewBuildPipeline(ref.Input().String(), input, c.OutputDocker, strategy, c.GetBuildEnvironment(environment), source); err != nil {
> -					return nil, fmt.Errorf("can't build %q: %v", ref.Input(), err)
> +

is it the same for --name? What kind of use of new-app /new-build produces
2 builds?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/5141/files#r42365386>.

rhcarvalho replied:
> It's illegal to specify --to when there are more than two builds defined - that check needs to be represented somewhere.

Added that check [here](https://github.com/openshift/origin/pull/5141/files#diff-67f42376bdba60c035c6268d160059baR894).

smarterclayton commented on pkg/generate/app/errors.go:
```diff
@@ -53,3 +55,13 @@ The argument %[1]q could apply to the following Docker images or OpenShift image
+}
+
+func (e CircularOutputReferenceError) Error() string {
+	return fmt.Sprintf("the input and output Docker image references are identical (%q), please specify a different output reference with --to", e.ImageReference)
```
--to is not generic to app builders - instead, newapp code needs to transform an error of type CircularOutputReferenceError into a message with `--to`.

rhcarvalho replied:
Done.

rhcarvalho commented:
@smarterclayton thanks for the review, I've updated code accordingly.
Still missing the check for "illegal to specify --to when there are more than two builds defined".

I don't know how to use 'oc new-app' / 'new-build' to create more than 1 build at once, except for doing 'oc new-app template.json'. Is there another way that I'm missing?

rhcarvalho commented:
Ok, I figured what I'm missing, working on it.

rhcarvalho commented:
Comments addressed, added similar checks for --to as there are for --name.

smarterclayton commented:
oc new-app openshift/php~<repo1> openshift/ruby~<repo2>

On Mon, Oct 19, 2015 at 9:20 AM, Rodolfo Carvalho <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> thanks for the
> review, I've updated code accordingly.
> Still missing the check for "illegal to specify --to when there are more
> than two builds defined".
>
> I don't know how to use 'oc new-app' / 'new-build' to create more than 1
> build at once, except for doing 'oc new-app template.json'. Is there
> another way that I'm missing?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/5141#issuecomment-149211289>.
>

rhcarvalho commented:
> oc new-app openshift/php~<repo1> openshift/ruby~<repo2>

Yes, I got that. Added similar checks as for `--name`.

rhcarvalho commented:
Rebased on top of newest kube godep, apparently this PR was not affected by the changes in kube.
@smarterclayton any other change you'd like to see here?

rhcarvalho commented:
@smarterclayton, based on your comments, my most recent push:
- added  unit test for [equality check of `DockerImageReference`](https://github.com/openshift/origin/pull/5141/files#diff-c7ce025f36a3834873271fcff866fdb4R910)
- moved the 3 methods to [`pipeline.go`](https://github.com/openshift/origin/pull/5141/files#diff-f26900778a8353c1e3b16d70b2dce230)
- [updated the help message for `--to`](https://github.com/openshift/origin/pull/5141/files#diff-077dc3559bcf0a05d791fc83d2d92197R75) ([I omitted the part between parenthesis in new-app](https://github.com/openshift/origin/pull/5141/files#diff-bd8a796157fcad2a1d7d9caa88fc6708R142) because there there's no `--to-docker` flag).
- turned [`validateOutputImageReference` into a function](https://github.com/openshift/origin/pull/5141/files#diff-67f42376bdba60c035c6268d160059baR465) (not a method on `AppConfig`). For consistency, I did the same with the validation for the `--name` argument, [`validateEnforcedName`](https://github.com/openshift/origin/pull/5141/files#diff-67f42376bdba60c035c6268d160059baR458).

PTAL

smarterclayton commented:
Looks good, you have some failing unit tests in jenkins.

rhcarvalho commented:
My bad, I was missing an import in pipeline_test.go when I refactored buildPipelines.
Should've been fixed by now.

---------------------------

BTW - how should I run unit tests for multiple packages (but not all)?

`hack/test-go.sh` works with [either one or no arguments](https://github.com/openshift/origin/blob/7c12c7ac80da0c4266d145b5fb6e62ade00967b4/hack/test-go.sh#L80-L86), so I was working around that by doing:
```
$ make check WHAT=pkg/generate/app TESTFLAGS="$(echo github.com/openshift/origin/pkg/{generate/app/cmd,image/api}) -v"
```

But this will proceed and test the latter two packages ignoring the build error of pkg/generate/app.
This catches the build error:
```
$ make check WHAT=. TESTFLAGS="$(echo github.com/openshift/origin/pkg/{generate/app{,/cmd},image/api}) -v"
```

Couldn't find another obvious way of testing multiple packages.

rhcarvalho commented:
Just opened https://github.com/openshift/origin/pull/5274 proposing a way to unit test multiple packages at once without hacks.

rhcarvalho commented:
@smarterclayton green tests, PTAL

rhcarvalho commented:
@smarterclayton, as a follow-up, we might want to do something to allow no build output.

I'll be on PTO next days and we might not catch each other online tomorrow because of the timezone difference...

> do you think we should have another argument like --no-output or a special value of --to?
> AFAIK with cobra/pflag there's no obvious way to determine if a flag was user provided or defaulted.

smarterclayton commented:
Yes I think we need to support that.  It would probably be another flag.

On Oct 21, 2015, at 11:48 AM, Rodolfo Carvalho <notifications@github.com>
wrote:

@smarterclayton <https://github.com/smarterclayton>, as a follow-up, we
might want to do something to allow no build output.

I'll be on PTO next days and we might not catch each other online tomorrow
because of the timezone difference...

do you think we should have another argument like --no-output or a special
value of --to?
AFAIK with cobra/pflag there's no obvious way to determine if a flag was
user provided or defaulted.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/5141#issuecomment-149939525>.

rhcarvalho commented:
> Yes I think we need to support that.  It would probably be another flag.

That's new functionality for new-app and new-build, can we do that as a follow-up? I don't want to block the fixes in this PR.

rhcarvalho commented:
Opened https://github.com/openshift/origin/issues/5296 to discuss and track the new-build/new-app without build output feature.

smarterclayton commented:
Needs a hack/test-cmd.sh set of testcased

rhcarvalho commented:
This PR was getting to big and out of control. I've rebased against master and kept just the commit to address #4980. 

#5072 and #4730 will be addressed in separate PRs.

@smarterclayton @bparees PTAL  [test]

rhcarvalho commented on pkg/generate/app/pipeline.go:
```diff
@@ -39,20 +39,32 @@ func NewImagePipeline(from string, image *ImageRef) (*Pipeline, error) {
-		Reference: image.DockerImageReference{
+	if len(to) > 0 {
+		var err error
+		outputImageRef, err = image.ParseDockerImageReference(to)
```
The pipeline output has a [`Reference`](https://github.com/openshift/origin/blob/aa8e6ba1e302ff7f5d89425d92b1e73c4979df0d/pkg/generate/app/imageref.go#L134) field of type `imageapi.DockerImageReference`, which [later on gets converted](https://github.com/openshift/origin/blob/aa8e6ba1e302ff7f5d89425d92b1e73c4979df0d/pkg/generate/app/imageref.go#L210-L229) to a `kapi.ObjectReference`, of `Kind` being either `ImageStreamTag` or `DockerImage`.

I believe we use a `imageapi.DockerImageReference` because it can, for what is needed here, represent both ImageStreamTag and DockerImage references, given that the format of ISTs is a subset of that of a DockerImage.

rhcarvalho commented on pkg/generate/app/cmd/newapp.go:
```diff
@@ -992,7 +1000,13 @@ func (c *AppConfig) run(acceptors app.Acceptors) (*AppResult, error) {
+	}
+
+	if len(c.To) > 0 {
+		if err := validateOutputImageReference(c.To); err != nil {
```
As I commented above, it can be either a DockerImage or a ImageStreamTag, depending on the `--to-docker` flag.

rhcarvalho commented on pkg/generate/app/cmd/newapp.go:
```diff
@@ -583,9 +584,16 @@ func (c *AppConfig) detectSource(repositories []*app.SourceRepository) error {
+
+func validateOutputImageReference(ref string) error {
+	if _, err := imageapi.ParseDockerImageReference(ref); err != nil {
+		return fmt.Errorf("invalid output image reference: %s", ref)
```
Yes, they should. The validation is done against DockerImage because that's a more general pattern. AFAIK, every IST is a parseable DockerImage ref.

What we could do is a more specific validation depending on `--to-docker`, go for that?
(such that `oc new-build <source> --to-docker=false --to repo:80/name/img:tag` would return an error instead of silently ignoring the "repo:80/" part).

rhcarvalho commented:
Still need to change the implementation to make new tests pass... WIP

rhcarvalho commented on test/cmd/builds.sh:
```diff
@@ -19,6 +19,18 @@ cat "${OS_ROOT}/Dockerfile" | oc new-build -D - --name=test
 oc new-build --dockerfile=$'FROM centos:7\nRUN yum install -y httpd'
 oc get bc/centos
+oc new-build -D $'FROM openshift/origin:v1.1' --to-docker
+[[ $(oc get bc/origin --template '{{with .spec.output.to}}{{.kind}} {{.name}}{{end}}') == "DockerImage library/origin:latest" ]]
```
@smarterclayton I would like your input here.

Currently, `oc new-build -D $'FROM openshift/origin:v1.1' --to-docker` configures the build output to a DockerImage "origin:latest".

Should we keep it that way? Or perhaps require --to to be specified.

smarterclayton replied:
I think that's ok for here

rhcarvalho commented:
@smarterclayton could you please have a look at the tests in `test/cmd/builds.sh`?
I'd like to validate that the expectations are correct.

rhcarvalho commented on test/cmd/builds.sh:
```diff
@@ -19,6 +19,18 @@ cat "${OS_ROOT}/Dockerfile" | oc new-build -D - --name=test
+oc new-build -D $'FROM openshift/origin:v1.1\nENV ok=1' --to origin-test:v1.1
+[[ $(oc get bc/origin-test --template '{{with .spec.output.to}}{{.kind}} {{.name}}{{end}}') == "ImageStreamTag origin-test:v1.1" ]]
+# Build from Dockerfile with output to DockerImage
+oc new-build -D $'FROM openshift/origin:v1.1\nENV ok=1' --to-docker --to openshift/origin:v1.1-test
```
This line is falling in Jenkins even though we delete all on line 24... maybe the delete just takes too long? Looking for ideas how to make this better.

bparees replied:
add a tryuntil that confirms the old buildconfig is gone.

rhcarvalho commented:
Maybe test-cmd is evaluating the "\n" in the test string...

php-coder commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -98,11 +98,12 @@ func NewCmdNewBuild(fullName string, f *clientcmd.Factory, in io.Reader, out io.
 	cmd.Flags().StringSliceVarP(&config.ImageStreams, "image", "i", config.ImageStreams, "Name of an image stream to to use as a builder.")
 	cmd.Flags().StringSliceVar(&config.DockerImages, "docker-image", config.DockerImages, "Name of a Docker image to use as a builder.")
 	cmd.Flags().StringVar(&config.Name, "name", "", "Set name to use for generated build artifacts")
+	cmd.Flags().StringVar(&config.To, "to", "", "Push built images to this image stream tag (or Docker image repository if --to-docker is set)")
```
I see that all other descriptions have dot at the end and this doesn't have it. May be add it?

BTW, I would name this option `--to-stream` to make name more obvious..

rhcarvalho replied:
> all other descriptions have dot at the end and this doesn't have it. May be add it?

Yes, well noted.

> BTW, I would name this option --to-stream to make name more obvious (IMHO).

The `--to` option can specify an image stream tag or a docker image reference:

`oc new-build <source> --to-docker --to my/docker:ref`
`oc new-build <source> --to my/image:stream`

rhcarvalho commented:
re[test]

rhcarvalho commented:
:green_heart: @smarterclayton @bparees PTAL

bparees commented:
lgtm.
[merge]

smarterclayton commented:
LGTM

rhcarvalho commented:
@bparees @smarterclayton I've updated [test/cmd/builds.sh](https://github.com/openshift/origin/pull/5141/files#diff-ba4d05695f474cd209c5a3c9a22de986) fixing the merge conflict with #6034, and updating the test code to use the wrapper methods.

rhcarvalho commented on test/cmd/builds.sh:
```diff
@@ -12,16 +12,64 @@ os::log::install_errexit
+
+os::cmd::expect_failure_and_text \
+	'oc new-build centos/ruby-22-centos7~https://github.com/openshift/ruby-ex centos/php-56-centos7~https://github.com/openshift/cakephp-ex --to invalid/argument' \
+	'error: only one component or source repository can be used when specifying an output image reference'
```
@smarterclayton @bparees also included this test to check "--to" given two source repos (invalid usage).

rhcarvalho commented:
Could anybody re-merge this, please?

liggitt commented on test/cmd/builds.sh:
```diff
@@ -12,16 +12,64 @@ os::log::install_errexit
 
+# expect_build_output_spec runs `oc new-build` with given arguments and expects
+# a build config with a given name to be created with a given build output spec
+function os::cmd::builds::expect_build_output_spec() {
```
Hmm, any way to use the test functions directly? Would like to avoid complex per-use bash functions. @stevekuznetsov any suggestions?

rhcarvalho replied:
The smart functions from the util are still used -- lines 26-28. This is to avoid repeating the same pattern over and over.

liggitt replied:
If it fails, what line gets logged? Can we tell which call into the shared function caused it?

stevekuznetsov replied:
The output will look like three tests are run, I think this should be pretty straight forward. However, no, the line file/line spec will show this function add the source, not the original calls.

stevekuznetsov replied:
I think an `os::cmd::expect_success` on the `new-build`, a function to build the correct template, and a `os::cmd::expect_success_and_text` on the `get` with the template should suffice - I'm not sure making the tests less explicit is worth it.

rhcarvalho replied:
I made it break:
```
[vagrant@openshiftdev origin]$ git diff
diff --git a/test/cmd/builds.sh b/test/cmd/builds.sh
index d1a10e7..038a611 100755
--- a/test/cmd/builds.sh
+++ b/test/cmd/builds.sh
@@ -39,7 +39,7 @@ os::cmd::expect_success 'oc get bc/test'
 os::cmd::builds::expect_build_output_spec \
        "--dockerfile=\$'FROM centos:7\nRUN yum install -y httpd'" \
        'centos' \
-       'ImageStreamTag centos:latest'
+       'ImageStreamTag centos:latest_wrong'
 
 # Build from Dockerfile with output to DockerImage
 os::cmd::builds::expect_build_output_spec \
```

And ran the tests:
```
[vagrant@openshiftdev origin]$ ./hack/test-cmd.sh test/cmd/builds.sh 
<...>
++ test/cmd/builds.sh
Now using project "default" from context named "default/127-0-0-1:28443/system:admin" on server "https://127.0.0.1:28443".
Now using project "cmd-builds" on server "https://127.0.0.1:28443".
SUCCESS: oc new-build openshift/ruby-20-centos7 https://github.com/openshift/ruby-hello-world.git
SUCCESS: oc get bc/ruby-hello-world
SUCCESS: cat "${OS_ROOT}/Dockerfile" | oc new-build -D - --name=test
SUCCESS: oc get bc/test
SUCCESS: oc delete all --all
SUCCESS: oc new-build --dockerfile=$'FROM centos:7\nRUN yum install -y httpd'
FAILURE: oc get bc/centos --template '{{with .spec.output.to}}{{.kind}} {{.name}}{{end}}': the output content test failed
Standard output from the command:
ImageStreamTag centos:latestThere was no error output from the command.
!!! Error in test/cmd/../../hack/cmd_util.sh:145
        'return 1' exited with status 1
Call stack:
        1: test/cmd/../../hack/cmd_util.sh:145 os::cmd::expect_success_and_text(...)
        2: test/cmd/builds.sh:28 os::cmd::builds::expect_build_output_spec(...)
        3: test/cmd/builds.sh:40 main(...)
Exiting with status 1
!!! Error in ./hack/test-cmd.sh:284
        '${test}' exited with status 1
Call stack:
        1: ./hack/test-cmd.sh:284 main(...)
Exiting with status 1
[FAIL] !!!!! Test Failed !!!!
<...>
```


With colors and all :+1: 


To compare, I made it break at a different place:
```
[vagrant@openshiftdev origin]$ git diff
diff --git a/test/cmd/builds.sh b/test/cmd/builds.sh
index d1a10e7..37effa1 100755
--- a/test/cmd/builds.sh
+++ b/test/cmd/builds.sh
@@ -33,7 +33,7 @@ function os::cmd::builds::expect_build_output_spec() {
 os::cmd::expect_success 'oc new-build openshift/ruby-20-centos7 https://github.com/openshift/ruby-hello-world.git'
 os::cmd::expect_success 'oc get bc/ruby-hello-world'
 os::cmd::expect_success 'cat "${OS_ROOT}/Dockerfile" | oc new-build -D - --name=test'
-os::cmd::expect_success 'oc get bc/test'
+os::cmd::expect_success 'oc get bc/test2'
 
 # Build from Dockerfile with output to ImageStreamTag
 os::cmd::builds::expect_build_output_spec \
```

And the test output:
```
[vagrant@openshiftdev origin]$ ./hack/test-cmd.sh test/cmd/builds.sh 
<...>
++ test/cmd/builds.sh
Now using project "default" from context named "default/127-0-0-1:28443/system:admin" on server "https://127.0.0.1:28443".
Now using project "cmd-builds" on server "https://127.0.0.1:28443".
SUCCESS: oc new-build openshift/ruby-20-centos7 https://github.com/openshift/ruby-hello-world.git
SUCCESS: oc get bc/ruby-hello-world
SUCCESS: cat "${OS_ROOT}/Dockerfile" | oc new-build -D - --name=test
FAILURE: oc get bc/test2: the command returned the wrong error code
There was no output from the command.
Standard error from the command:
Error from server: buildconfig "test2" not found
!!! Error in test/cmd/../../hack/cmd_util.sh:145
        'return 1' exited with status 1
Call stack:
        1: test/cmd/../../hack/cmd_util.sh:145 os::cmd::expect_success(...)
        2: test/cmd/builds.sh:36 main(...)
Exiting with status 1
!!! Error in ./hack/test-cmd.sh:284
        '${test}' exited with status 1
Call stack:
        1: ./hack/test-cmd.sh:284 main(...)
Exiting with status 1
[FAIL] !!!!! Test Failed !!!!
<...>
```

In both cases we get a reasonable stack trace, a good indication of which line failed, and the exact command that failed. Agree?

stevekuznetsov replied:
Very soon (next week?) I will be tweaking the output of the functions to read `(Running|SUCCESS|FAILURE): file:line_number: wrapper args`, where `file:line_number:` will be the place where the wrapper is called. For this case, the failure text will bring you to the call inside the function when instead you want the call of `os::cmd::builds::expect_build_output_spec`. You could still see that from the stack trace, but I'm not sure if it's worth the confusion - especially since jUnit integration will use that line number and file to track flakes, we won't get useful data for flakes using this function.

stevekuznetsov replied:
Also, small nit - this function should live in `hack/builds.sh` once https://github.com/openshift/origin/issues/5844 is sorted out, and maybe in `hack/cmd_util.sh` or `hack/util.sh` for now, not here.

stevekuznetsov replied:
```sh
template='{{with .spec.output.to}}{{.kind}} {{.name}}{{end}}'
os::cmd::expect_success 'oc new-build <args>'
os::cmd::expect_success_and_text "oc get bc/<name> --template '${template}'" '<text>'
```

This is fairly concise, clear, and doesn't have drawbacks for jUnit. I think it's OK if our tests are verbose, literal, and the less bash trickery we use the better.

rhcarvalho replied:
Just curiosity, why are we creating a _bash_ test framework?

liggitt replied:
Testing cli invocations is (somewhat) reasonably done via bash. We already had a bash framework, we're just trying to make it (more) reasonable to jump in and know exactly what is supposed to be getting checked.

rhcarvalho commented:
Fixed mixing spaces and tabs in `test/cmd/builds.sh`.

stevekuznetsov commented on test/cmd/builds.sh:
```diff
@@ -12,16 +12,64 @@ os::log::install_errexit
+	local template='{{with .spec.output.to}}{{.kind}} {{.name}}{{end}}'
+	local get_output_spec_cmd="oc get bc/${expected_buildconfig_name} --template '${template}'"
+
+	os::cmd::expect_success 'oc delete all --all'
```
This line looks like something this function shouldn't be doing - I wouldn't expect this side effect from the name.

liggitt replied:
I dunno, I would rather see repeated commands than have to manage another level of indirection. Do we actually need to do the delete all every time if we use different names?

stevekuznetsov replied:
> I would rather see repeated commands

I agree. Let test files be verbose.

stevekuznetsov commented on test/cmd/builds.sh:
```diff
@@ -12,16 +12,64 @@ os::log::install_errexit
+# expect_build_output_spec runs `oc new-build` with given arguments and expects
+# a build config with a given name to be created with a given build output spec
+function os::cmd::builds::expect_build_output_spec() {
+	if [[ $# -ne 3 ]]; then echo "os::cmd::builds::expect_build_output_spec expects three argumments, got $#"; fi
```
`s/argumments/arguments/` - originally my typo, sorry

php-coder commented on test/cmd/builds.sh:
```diff
@@ -12,16 +12,64 @@ os::log::install_errexit
 
+# Build from Dockerfile with output to ImageStreamTag
+os::cmd::builds::expect_build_output_spec \
+	"--dockerfile=\$'FROM centos:7\nRUN yum install -y httpd'" \
```
BTW, is there any reason why we use `--dockerfile` here and `-D` in other cases?

rhcarvalho replied:
I kept the original test using "--dockerfile" while using "-D" for the others so that we have both short and long flag name in tests.

liggitt commented on test/cmd/builds.sh:
```diff
@@ -12,16 +12,64 @@ os::log::install_errexit
+	os::cmd::expect_success 'oc delete all --all'
+	os::cmd::expect_success "oc new-build ${newbuild_args}"
+	os::cmd::expect_success_and_text "${get_output_spec_cmd}" "^${expected_output_spec}\$"
+}
```
The template var could be reused, the `delete all` I would only expect once, and it's easier to follow the test if the top level invokes new-build followed by an output check, rather than a wrapper function

rhcarvalho replied:
The `delete all` (or at least `delete is`) is needed most if not all times between those tests, because they all generate the same name for image streams.

rhcarvalho commented:
@stevekuznetsov @liggitt I've flattened out the tests PTAL

stevekuznetsov commented on test/cmd/builds.sh:
```diff
@@ -18,10 +18,35 @@ os::cmd::expect_success 'oc new-build openshift/ruby-20-centos7 https://github.c
+os::cmd::expect_success_and_text "oc get bc/origin --template '${template}'" '^DockerImage library/origin:latest$'
+
+# Build from Dockerfile with given output ImageStreamTag spec
+os::cmd::expect_success 'oc delete is/origin'
```
If the `oc delete` lines are cleanup from the previous test, I'd stick them there instead of them leading the next test. Otherwise these LGTM as far as formatting and readability, can't comment on correctness.

liggitt commented:
[test]

bparees commented:
lgtm [merge]
